src/constants.rs

```diff
@@ -1,4 +1,333 @@
+        // For the last window, we compute [k * (8^w) - sum]B, where sum is defined
+        // as sum = \sum_{j = 0}^{`num_windows - 2`} 8^j
+        let sum = (0..(num_windows - 1)).fold(C::ScalarExt::zero(), |acc, w| {
+            acc + C::ScalarExt::from_u64(H as u64).pow(&[w as u64, 0, 0, 0])
+        });
+        window_table.push(
+            (0..H)
+                .map(|k| {
+                    // scalar = k * (8^w) - sum, where w = `num_windows - 1`
+                    let scalar = C::ScalarExt::from_u64(k as u64)
+                        * C::ScalarExt::from_u64(H as u64).pow(&[
+                            (num_windows - 1) as u64,
+                            0,
+                            0,
+                            0,
+                        ])
+                        - sum;
+                    (self.0 * scalar).to_affine()
+                })
+                .enumerate()
+                .fold([C::identity(); H], |mut window, (index, entry)| {
+                    window[index] = entry;
+                    window
+                }),
+        );
+
+        window_table
+    }
+
+    fn compute_lagrange_coeffs(&self, num_windows: usize) -> Vec<[C::Base; 8]> {
+        // We are interpolating over the 3-bit window, k \in [0..8)
+        let points: Vec<_> = (0..H).map(|i| C::Base::from_u64(i as u64)).collect();
+
+        let window_table = self.compute_window_table(num_windows);
+
+        window_table
+            .iter()
+            .map(|window_points| {
+                let x_window_points: Vec<_> = window_points
+                    .iter()
+                    .map(|point| point.get_xy().unwrap().0)
+                    .collect();
+                lagrange_interpolate(&points, &x_window_points)
+                    .iter()
+                    .enumerate()
+                    .fold([C::Base::default(); H], |mut window, (index, entry)| {
+                        window[index] = *entry;
+                        window
+                    })
+            })
+            .collect()
+    }
+
+    /// For each window, z is a field element such that for each point (x, y) in the window:
+    /// - z + y = u^2 (some square in the field); and
+    /// - z - y is not a square.
+    fn find_zs(&self, num_windows: usize) -> Option<Vec<u64>> {
+        // Closure to find z for one window
+        let find_z = |window_points: &[C]| {
+            assert_eq!(H, window_points.len());
+
+            let ys: Vec<_> = window_points
+                .iter()
+                .map(|point| point.get_xy().unwrap().1)
+                .collect();
+            let z_for_single_y = |y: C::Base, z: u64| {
+                let sum_y_is_square: bool = (y + C::Base::from_u64(z)).sqrt().is_some().into();
+                let sum_neg_y_is_square: bool = (-y + C::Base::from_u64(z)).sqrt().is_some().into();
+                (sum_y_is_square && !sum_neg_y_is_square) as usize
+            };
```

Comment:
We probably want to remember u, so that we can use it in the witness generation (rather than having to recompute the square root).